### PR TITLE
Add End2End tests

### PR DESCRIPTION
**Pull request by JanusWu**

### CODE REVIEW

**Comment by JanusWu** [comment]
Here are something I want you know:
1. There is a build failure in CosmosDBJava.java under End2End\functionannotation
`[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /E:/JavaTest/Test0626/azure-functions-java-worker/src/test/End2End/functionannotation/src/main/java/com/function/CosmosDBJava.java:[14,14] cannot find symbol
  symbol:   class DocumentDBInput
  location: class com.function.CosmosDBJava`

2. According to [#171](https://github.com/Microsoft/azure-maven-plugins/pull/171), the dependency `azure-functions-java-core` should be updated to `azure-functions-java-library`. But the code haven't updated in this PR.

**Comment by pragnagopa** [comment]
Delete documentDB. There is only cosmosdb in Functions v2 https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-cosmosdb-v2

**Comment by JanusWu** [comment]
@pragnagopa , Is there any other problem in this PR?

File: src/test/End2End/functest/pom.xml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  </properties>
+  <dependencies>
+  <dependency>
+      <groupId>com.microsoft.azure</groupId>
+      <artifactId>azure-documentdb</artifactId>
+      <version>1.15.1</version>
+    </dependency>
+    <dependency>
+      <groupId>com.microsoft.azure</groupId>
+      <artifactId>azure-functions-java-core</artifactId>
```
Comment (pragnagopa) [inline src/test/End2End/functest/pom.xml]: Azure-functions-library

File: src/test/End2End/functest/pom.xml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  <groupId>com.microsoft.serverless</groupId>
+  <artifactId>azure-functions-java-sample</artifactId>
+  <version>1.0-SNAPSHOT</version>
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+  </properties>
+  <dependencies>
+  <dependency>
+      <groupId>com.microsoft.azure</groupId>
+      <artifactId>azure-documentdb</artifactId>
```
Comment (pragnagopa) [inline src/test/End2End/functest/pom.xml]: Delete

File: src/test/End2End/functest/src/main/functions/cosmosDBOut/function.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{
+    "scriptFile": "../../../../target/azure-functions-java-sample-1.0-SNAPSHOT.jar",
+    "entryPoint": "com.microsoft.serverless.CosmosDBTests.cosmosDBOut",
+    "bindings": [
```
Comment (pragnagopa) [inline src/test/End2End/functest/src/main/functions/cosmosDBOut/function.json]: Remove seeverless from package name everywhere

File: src/test/End2End/functest/pom.xml [hunk 4]
```diff
@@ -0,0 +1,58 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>com.microsoft.serverless</groupId>
+  <artifactId>azure-functions-java-sample</artifactId>
```
Comment (pragnagopa) [inline src/test/End2End/functest/pom.xml]: Groupid: com.microsoft.azure.functions
Artifactid: azure-functions-EndToEndTests

File: src/test/End2End/functest/pom.xml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        </filter>
+    </filters>
+    <!-- Additional configuration. -->
+</configuration>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+</project>
```
Comment (pragnagopa) [inline src/test/End2End/functest/pom.xml]: Add readme in EndToEndTests folder with instructions on how to run tests

**Comment by pragnagopa** [comment]
@JanusWu - Ping!

**Comment by pragnagopa** [comment]
Add EventGridTrigger tests. More info on eventgrid https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-event-grid

**Comment by JanusWu** [comment]
Sorry for the delay of response. I have made much changes on the code and I'm testing them now. I will list the status for each of these tests after that.

**Comment by JanusWu** [comment]
As for the EventGridTrigger tests, I will add them later.

File: End2End/functest/src/main/functions/blobTriggerTest/function.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{
+    "scriptFile": "../../../../target/azure-functions-java-sample-1.0-SNAPSHOT.jar",
+    "entryPoint": "com.microsoft.serverless.BlobTests.blobTriggerTest",
```
Comment (pragnagopa) [inline End2End/functest/src/main/functions/blobTriggerTest/function.json]: remove serverless

File: End2End/functest/src/main/functions/httpTriggerJava/function.json [hunk 7]
```diff
@@ -0,0 +1,18 @@
+{
+    "scriptFile": "../../../../target/azure-functions-java-sample-1.0-SNAPSHOT.jar",
+    "entryPoint": "com.microsoft.serverless.HttpTests.httpTriggerJava",
+    "bindings": [
```
Comment (pragnagopa) [inline End2End/functest/src/main/functions/httpTriggerJava/function.json]: dup of above?

File: End2End/functest/src/main/functions/queueTriggerTest/function.json [hunk 8]
```diff
@@ -0,0 +1,13 @@
+{
+    "scriptFile": "../../../../target/azure-functions-java-sample-1.0-SNAPSHOT.jar",
+    "entryPoint": "com.microsoft.serverless.QueueTests.queueTriggerTest",
+    "bindings": [
```
Comment (pragnagopa) [inline End2End/functest/src/main/functions/queueTriggerTest/function.json]: Same here..what is the difference between queueTrigger and queueTriggerTest

File: End2End/functest/src/main/functions/queueTriggerTest/function.json [hunk 9]
```diff
@@ -0,0 +1,13 @@
+{
+    "scriptFile": "../../../../target/azure-functions-java-sample-1.0-SNAPSHOT.jar",
+    "entryPoint": "com.microsoft.serverless.QueueTests.queueTriggerTest",
```
Comment (pragnagopa) [inline End2End/functest/src/main/functions/queueTriggerTest/function.json]: remove serverless

File: End2End/functest/src/main/java/com/microsoft/azure/functions/Blob.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
+package com.microsoft.azure.functions;
+
```
Comment (pragnagopa) [inline End2End/functest/src/main/java/com/microsoft/azure/functions/Blob.java]: update pacakge name to com.microsoft.azure.functions.EndToEnd

File: End2End/functest/src/main/functions/queueTriggerTest/function.json (hunk 8 above)
Comment (JanusWu) [reply End2End/functest/src/main/functions/queueTriggerTest/function.json]: Sorry for the dupe like queueTrigger and queueTriggerTest. The folders contain "*Test" should have 
been deleted, but seems they were left behind with my careless mistake.

File: End2End/functionannotation/src/main/java/com/microsoft/azure/functions/Blob.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
+package com.microsoft.azure.functions;
+
+import com.microsoft.azure.functions.annotation.*;
+
+/**
+ * Azure Functions with Azure Blob.
+ */
+public class Blob {
+    /**
```
Comment (pragnagopa) [inline End2End/functionannotation/src/main/java/com/microsoft/azure/functions/Blob.java]: Instead having these tests under functionannotation folder, can you please combine the tests with [endtoend/Blob.java](https://github.com/JanusWu/azure-functions-java-worker/blob/f84020e0c669e505dd1ede322827b2062da3e8c0/End2End/functest/src/main/java/com/microsoft/azure/functions/endtoend/Blob.java)

File: End2End/functionannotation/src/main/java/com/microsoft/azure/functions/Blob.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        return request.getBody().length + "bytes";
+    }
+    
+    /**
+    * This function will be invoked when a new or updated blob is detected at the specified path. The blob contents metatdata are provided as input to this function.
+    */
+    // @FunctionName("BlobMetaDataJava")
+    // @StorageAccount("AzureWebJobsStorage")
+    // public void blobMetaData(
+    //         @BlobTrigger(name = "content", path = "samples/{fileName}.{fileExtension}", dataType = "binary") byte[] content,
```
Comment (pragnagopa) [inline End2End/functionannotation/src/main/java/com/microsoft/azure/functions/Blob.java]: Delete commented code if not needed

File: End2End/functionannotation/src/main/java/com/microsoft/azure/functions/CosmosDB.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
+package com.microsoft.azure.functions;
+
+import com.microsoft.azure.functions.annotation.*;
+import java.util.*;
+
+public class CosmosDB {
+    /**
+    * This function will be invoked when a message add to the queue.The documentDB use as input to get the item with the name of the message
```
Comment (pragnagopa) [inline End2End/functionannotation/src/main/java/com/microsoft/azure/functions/CosmosDB.java]: Same goes for all the tests under functionannotation folder. Combine tests with EndToEnd and delete functionannotation folder.

File: End2End/README.md [hunk 14]
```diff
@@ -0,0 +1,13 @@
+# Microsoft Azure Functions in Java
+
+This repo contains several Java Azure Functions annotation samples and without annotation samples for different events.
```
Comment (pragnagopa) [inline End2End/README.md]: This folder contains EndToEnd Java Azure function tests.

File: End2End/README.md [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+
+## Prerequisites
+- JDK 1.8
+- Maven 3.0+
+- [.NET Core SDK](https://www.microsoft.com/net/learn/get-started/windows)
+- [Azure Functions Core Tools v2](https://docs.microsoft.com/en-us/azure/azure-functions/functions-run-local#v2)
```
Comment (pragnagopa) [inline End2End/README.md]: Azure subscription is required right?

File: End2End/README.md [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+- [Maven Plugin for Azure Functions](https://github.com/Microsoft/azure-maven-plugins/blob/develop/azure-functions-maven-plugin/README.md)
+
+## Azure Functions List and Test Pass
```
Comment (pragnagopa) [inline End2End/README.md]: Add instructions on how to run the tests. For example, login to azure and manually create the resources needed for the tests like storage account etc.
- Please list all the azure resources that need to be created
- Please list all the environment variables that need to be set to run them locally.

File: End2End/README.md (hunk 15 above)
Comment (JanusWu) [reply End2End/README.md]: That's right.

File: End2End/README.md (hunk 16 above)
Comment (JanusWu) [reply End2End/README.md]: Yes, I will list that.

**Comment by JanusWu** [comment]
Yes, all tests under folder functionannotaion have corresponding tests under functest folder. The difference between these two kind of tests is that functions under functionannotaion have annotations, functions under functest don't have annotations but have corresponding function.json files.

If we only want to keep one kind of tests left，here are two approaches：
1. Remain tests under functest folder, tests under functionannotation folder could be all deleted, nothing need to be combined.
2. Remain tests under functionannotation folder, but update it's package structure to `com.microsoft.azure.functions.endtoend` like [EndToEnd-Java-Azure-Function-Tests](https://github.com/JanusWu/EndToEnd-Java-Azure-Function-Tests)

Which approach do you prefer? Or let me know if I misunderstand your opinion.

File: End2End/functionannotation/local.settings.json [hunk 17]
```diff
@@ -0,0 +1,7 @@
+{
+  "IsEncrypted": false,
+  "Values": {
```
Comment (pragnagopa) [inline End2End/functionannotation/local.settings.json]: Delete this file. It contains secrets

**Comment by pragnagopa** [comment]
Ok. Thanks for explaining.  Let's keep the tests. 
- Rename funcitonannocation to mavenPluginTests
- Add readme in that folder with the setup needed
- Add a scripts with maven commands that generate function.json and deploy to azure
  -  Add .bat file for windows
  - Add .sh file for mac/linux

File: End2End/functionannotation/pom.xml [hunk 18]
```diff
@@ -0,0 +1,138 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>com.microsoft.azure.functions</groupId>
```
Comment (pragnagopa) [inline End2End/functionannotation/pom.xml]: groupId: com.microsoft.azure.functions.endtoend

File: End2End/functionannotation/pom.xml [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+    <artifactId>azure-functions-EndToEndTests</artifactId>
```
Comment (pragnagopa) [inline End2End/functionannotation/pom.xml]: artificatId: azure-functions-mavenPluginTests

**Comment by pragnagopa** [comment]
@jdneo  - Just FYI @JanusWu will be sending a PR to add EndToEnd tests for maven plugin.

**Comment by JanusWu** [comment]
I have sent a PR [#265](https://github.com/Microsoft/azure-maven-plugins/pull/265) to repo azure-maven-plugins to add fucntionannotation tests. For the reset of code under functest folder and README, I will update them in a couple of days.

**Comment by JanusWu** [comment]
Close this PR and I will open a new one.

**Pull request closed without merging**